### ISSUE TITLE
Request for Timeline and Plans to Open-Source Code

### ISSUE THREAD

**Issue by JackeyWang777** [heyzude/BMTPT#1]
Hi,

I am a student who is highly interested in your project. I have noticed its great potential and I am particularly intrigued by the code implementation. I would like to inquire about any plans you may have to open-source the code for this project.

If there are plans to open-source the project, I would like to know when the source code will be made accessible. This will allow me to better understand and learn from the implementation details of the project, as well as provide an opportunity to contribute to its development.

If you haven't considered open-sourcing the project yet, I strongly encourage you to explore this possibility. Open-sourcing the code can attract more developers to participate in the project, driving its progress and benefiting the entire developer community.

I am eagerly looking forward to hearing from you about the plans for open-sourcing the code. If there are any further discussions or opportunities for contribution, please let me know. Thank you very much for your time and effort!

Best regards

**Comment by heyzude** [comment]
Hi, and thank you very much for your interest in our work!

We apologize for the delay in the code release. 
We plan to release the refactored code, probably by late January. However, In case we can't meet the timeline we will release the original code to ensure transparency and accessibility.

**Comment by heyzude** [comment]
Before fully releasing the code, I want to share the core implementation part of our work. I think this can give you a rough picture of the implementation.

The below picture describes the batch composition during source posterior training:
<img width="579" alt="image" src="https://github.com/heyzude/BMTPT/assets/12981081/0a6d8d22-c25b-45be-99e1-14e40f55ad49">

## Core code blocks
### Source posterior learning
```
    def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor, Any]]) -> torch.Tensor:

        model.train()
        inputs = self._prepare_inputs(inputs)

        # Refer to Chapter 4.3 of our paper
        for k in inputs.keys():
            if k == 'task':
                inputs[k] = inputs[k] * model.num_particles
            else:
                inputs[k] = inputs[k].repeat(model.num_particles, 1)
        batch_size = inputs['input_ids'].size(0)

        assert not is_sagemaker_mp_enabled() 
        with torch.enable_grad():
            particles = model.particles
            particles_copied = particles.repeat_interleave(batch_size//model.num_particles, dim=0)

            if self.use_amp:
                with autocast():
                    loss = self.compute_loss(model=model, inputs=inputs, particles=particles_copied)
            else:
                loss = self.compute_loss(model=model, inputs=inputs, particles=particles_copied)
            
            if self.args.n_gpu > 1:
                loss = loss.mean()  # mean() to average on multi-gpu parallel training
            if self.args.gradient_accumulation_steps > 1 and not self.deepspeed:
                # deepspeed handles loss scaling by gradient_accumulation_steps in its `backward`
                loss = loss / self.args.gradient_accumulation_steps
            
            padding_mask = inputs['labels'].eq(-100)
            num_active_elements = padding_mask.numel() - padding_mask.long().sum()
            llik = -loss * num_active_elements
            lpost = llik

            grads = autograd.grad(lpost, particles_copied, only_inputs=True, allow_unused=True)[0]
            aggregated_grads = torch.stack([grads[j * self.args.per_device_train_batch_size : (j+1) * self.args.per_device_train_batch_size].sum(0) for j in range(model.num_particles)])

            # Standard SVGD
            # K_XX = model.kernel(particles, particles.detach())
            # s1_kernel = K_XX.clone().detach()
            # s2 = -torch.autograd.grad(K_XX.sum(), particles)[0]
            # phi = s1_kernel.matmul(aggregated_grads) + s2
            # phi /= model.num_particles
            # phi = -phi

            # Damped SVGD (for alleviating the variance underestimation problem in standard SVGD)
            K_XX = model.kernel(particles, particles.detach())
            same_indices = K_XX == model.kernel.f(0.)

            kernel_without_same = K_XX.clone()
            kernel_without_same[same_indices] = 0.
            s1_kernel = kernel_without_same.clone().detach()
            s2 = -torch.autograd.grad(kernel_without_same.sum(), particles)[0]

            s1_ones = torch.zeros(model.num_particles, model.num_particles)
            s1_ones[same_indices] = 1.
            pdb.set_trace()
            phi = s1_kernel.matmul(aggregated_grads) + s2 + min(1., (model.kernel.f(1.) - 5 * (model.num_particles / particles[0].size(0)) * model.kernel.f_div(1.)) / model.kernel.f(0.)) * (s1_ones.cuda().matmul(aggregated_grads))  
            phi /= model.num_particles
            phi = -phi

            model.particles.grad = phi

            return 0
```
This code block implements Damped SVGD (https://openreview.net/pdf?id=Qycd9j5Qp9J) to learn the prompt posterior from source tasks. In the code, the repeated input (by .repeat)`inputs`, and the repeated source prompts (particles; by .repeat_interleave) `particles_copied` are concatenated at `self.compute_loss` function that is not shown here. Since the cross-entropy loss (log p (y|x, \theta)) can be interpreted as posterior under uniform prior (as in the paper), `self.compute_loss` is actually the posterior in the SVGD equation (the loss is then converted to log likelihood and the log likelihood becomes log posterior as they're proportional, under uniform prior assumption). After getting the gradients for each entry of the batch (batchsize: M x K), the gradients for each particle is from `grads` are extracted and aggregated into `aggregated_grads` (shape: M x particle_dimension). Then (Damped) SVGD is calculated.

### 2. Target adaptation
```
loss = loss + (0.5 * (model.prompt_config.sigma_value**(-2))) * torch.norm(target_prompt - source_prompts.mean(0))**2
```
This code implements the target adaptation loss which is stated in Chapter 4.1 (eq. 6)
The second term of this loss can be thought of as just a `log`ed Gaussian distribution, centered around the mean of particles obtained from source posterior learning step.
The intention behind this is to bring more information about the source tasks by transferring the source posterior to target adaptation prior.
